docs/tools.md

```diff
@@ -0,0 +1,101 @@
+# Tools
+
+## Introduction
+
+Tools in the context of an agent refer to additional functionalities or capabilities integrated with the agent to perform specific tasks beyond text processing.
+
+These tools extend the agent's abilities, allowing it to interact with external systems, access information, and execute actions.
+
+## Development
+
+### Writing a new tool
+
+To create a tool, the `Tool` or `DynamicTool` class must be implemented from [tools base](../src/tools/base.ts). When starting to write tools, it is recommended to implement the `Tool` class. The `DynamicTool` class is an extension of the `Tool` class that allows for dynamic forms of input. Refer to the list of [examples](#examples) for some simple examples or any of the built-in tools in order to guide the creation of new tools.
+
+Tools MUST do the following:
+
+- Implement the `Tool` class:
+
+  ```typescript
+  export class MyNewTool extends Tool<MyNewToolOutput, ToolOptions, ToolRunOptions> {
+    // tool implementation
+  }
+  ```
+
+- Be given a unique name:
+
+  ```typescript
+  name = "MyNewTool";
+  ```
+
+- Provide a natural language description of what the tool does:
+
+  ❗Important: this description is used by the agent to determine when the tool should be used. It's probably the most important aspect of your tool and you should experiment with different natural language descriptions to ensure the tool is used in the correct circumstances.
+
+  ```typescript
+  description = "Takes X action when given Y input resulting in Z output";
+  ```
+
+- Declare an input schema:
+
+  This is used to define the format of the input to your tool. The agent will formalise the natural language input(s) it has received and structure them into the fields described in the tool's input. The input schema is specified using [Zod](https://github.com/colinhacks/zod).
+
+  ```typescript
+  // any Zod definition is good here, this is typical simple example
+  inputSchema = z.object({
+    // list of key-value pairs
+  });
+  ```
+
+- Implement initialisation:
+
+  The unnamed static block is executed when your tool is called for the first time. It is used for registering your tool to the agent and any other custom initialisation that may be required.
```

Comment:
It's mainly used to automatically register the tool as `serializable`. If omitted, serialization may or may not occur.
See: https://github.com/i-am-bee/bee-agent-framework/blob/main/src/internals/serializable.ts#L41-L43

Reply:
Resolved (pending further review) in 9f58a2b